### PR TITLE
avoiding division by 0 in slicing

### PR DESCRIPTION
**Pull request by ipdemes**

### CODE REVIEW

**Comment by magnatelee** [comment]
I feel we shouldn't be in this code path if the tile shape is empty. can you instead add a condition that the store is empty to this if statement? https://github.com/nv-legate/legate.core/blob/branch-22.10/legate/core/store.py#L1278 that will prevent a tile partition from ever being created for an empty store.

**Comment by ipdemes** [comment]
@magnatelee : I have tried adding `or self.size ==0` (adding logic for size method), but this didn't help. I actually want slicing to happen (I call project on (0,1) shape)

**Comment by magnatelee** [comment]
@ipdemes if that's the case, then I think you should instead change how the slicing is done. any projection on an empty store can reuse the storage of the original store, instead of creating a sub-store.

**Comment by ipdemes** [comment]
@magnatelee OK. I tried to reuse storage when size==0 in my latest commit. Please have a look

File: legate/core/store.py [hunk 1]
```diff
@@ -950,6 +950,15 @@ def ndim(self) -> int:
...
             return self._shape.ndim
 
+    @property
+    def size(self) -> int:
+        s = 1
+        if self._ndim == 0:
+            return s
+        for p in self.shape:
+            s *= p
+        return s
```
Comment (bryevdv) [inline legate/core/store.py:960]: FYI `math.prod` was added in 3.8 
```
In [2]: math.prod((1,2,3))
Out[2]: 6

In [3]: math.prod(())
Out[3]: 1
```
(assuming shape is empty when ndim is zero)

File: legate/core/store.py [hunk 2]
```diff
@@ -727,10 +727,10 @@ def slice(self, tile_shape: Shape, offsets: Shape) -> Storage:
         # region in all dimensions of the parent region, then we'll make
         # a disjoint tiled partition with as many children as possible
         shape = self.get_root().extents
+
```
Comment (magnatelee) [inline legate/core/store.py:730]: please revert this bogus change

File: legate/core/store.py [hunk 3]
```diff
@@ -727,10 +727,10 @@ def slice(self, tile_shape: Shape, offsets: Shape) -> Storage:
...
         can_tile_completely = (offsets % tile_shape).sum() == 0 and (
             shape % tile_shape
         ).sum() == 0
-
```
Comment (magnatelee) [inline legate/core/store.py:733]: please revert this bogus change

File: legate/core/store.py (hunk 1 above)
Comment (magnatelee) [reply legate/core/store.py:960]: this is the code from when the minimum python version was 3.7. now we drop python 3.7 from the list of supported versions, we can and should replace this with `math.prod`. (and I suspect this isn't the only place where we roll a custom `prod` implementation.)

File: legate/core/store.py [hunk 4]
```diff
@@ -1119,6 +1131,10 @@ def slice(self, dim: int, sl: slice) -> Store:
             )
 
         size = self.shape[dim]
+
+        if size == 0 and sl == slice(None):
```
Comment (magnatelee) [inline legate/core/store.py:1135]: I think we can change this to `if size == 0 or sl == slice(None):`, as a slicing is no-op with either of them.

FYI, it looks like NumPy allows slice indices to be out-of-bounds, which would end up resulting an empty slice, whereas we raise an exception below. This is something we should fix after this PR is merged.

File: legate/core/store.py (hunk 1 above)
Comment (ipdemes) [reply legate/core/store.py]: Yes, this was basically copy-paste from cunumeirc. I have fixed it with the latest commit

File: legate/core/store.py (hunk 4 above)
Comment (ipdemes) [reply legate/core/store.py]: makes sense.

**Review by magnatelee — APPROVED** [review]

**Pull request merged**